### PR TITLE
feat(term): pipe osc and dcs commands to underlying terminal emulator

### PR DESCRIPTION
**Pull request by sassanh**
Cursor shape changes are working but `imgcat` of iTerm2 still doesn't work, I guess Neovim/libuv/ncurses library/something is changing terminal mode/settings. I don't have much knowledge about that but if I write the exact same bytes to a file and `tail -f` that file in another tab of iTerm2, image shows correctly, but in Neovim it shows like this:

image

### CODE REVIEW

**Comment by zeertzjq** [comment]
This is probably not wanted for remote UIs. Maybe the approach in #11290 is better?

**Comment by sassanh** [comment]
Thanks for the direction! I followed the approach in [#11290](https://github.com/neovim/neovim/pull/11290) and it resolved the imgcat issue:
image
Of course libvterm doesn't know anything about the size of that image and therefor gets confused where to render next line, but I think this is not important at all at this stage and one can easily create an image viewer plugin for Neovim and take care of this in its own plugin (write new lines, etc).

One problem I noticed with this new approach is rendering hyperlinks explained [here](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda)

This command doesn't render a hyperlink in iTerm2:
```bash
printf "\e]8;;https://neovim.io\e\\A link to Neovim\e]8;;\e\\"
```
while this one does:
```bash
printf "\e]8;;https://neovim.io\e\\A link to Neovim"; sleep .01; printf "\e]8;;\e\\ \n"
```
So I'm thinking with the new approach, maybe we are rendering all osc commands sent in a small time window together and then write "A link to Neovim" in the terminal. Or maybe the opposite (we first write "A link to Neovim" and then osc commands). Is there a way to flush the output buffer after initial osc command and before "A link to Neovim" is written as well as after when it has been written and before the second osc command and make sure it actually gets written in between? I tried this with no luck:
```diff
diff --git a/src/nvim/terminal.c b/src/nvim/terminal.c
index 0e6c540bc..3f920710f 100644
--- a/src/nvim/terminal.c
+++ b/src/nvim/terminal.c
@@ -842,6 +842,7 @@ bool terminal_running(const Terminal *term)
 static int term_unhandled_dcs(const char *command, size_t command_length, VTermStringFragment fragment, void *user)
 {
   if (fragment.initial) {
+    ui_call_flush();
     char *buffer = xmalloc(command_length + 3);
     snprintf(buffer, command_length + 3, "\033P%s;", command);
     ui_call_term_unhandled((String){.data = buffer, .size = command_length + 5});
@@ -853,6 +854,7 @@ static int term_unhandled_dcs(const char *command, size_t command_length, VTermS
   }
   if (fragment.final) {
     ui_call_term_unhandled((String){.data = "\033\\", .size = 2});
+    ui_call_flush();
   }
   return 1;
 }
@@ -860,6 +862,7 @@ static int term_unhandled_dcs(const char *command, size_t command_length, VTermS
 static int term_unhandled_osc(int command, VTermStringFragment fragment, void *user)
 {
   if (fragment.initial) {
+    ui_call_flush();
     char *buffer = xmalloc((size_t)((ceil(log10(command))+1)*sizeof(char)));
     int len = sprintf(buffer, "\033]%d;", command);
     ui_call_term_unhandled((String){.data = buffer, .size = (size_t)len});
@@ -871,6 +874,7 @@ static int term_unhandled_osc(int command, VTermStringFragment fragment, void *u
   }
   if (fragment.final) {
     ui_call_term_unhandled((String){.data = "\033\\", .size = 2});
+    ui_call_flush();
   }
   return 1;
 }
```

**Comment by sassanh** [comment]
I removed the virtual cursor in tui to show the real cursor. Now lots of tests are failing because they expect the `{reverse = true}` attribute to show the cursor position (`{1: }` most of the time) while it has changed to `^` (I think this is the real cursor indicator) now.
So lots of snapshots need to be updated, my question is: do we have a mechanism like jest --update-snapshots or should it be done manually for each test?

**Comment by zeertzjq** [comment]
Conflicting PR: #22159

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -1429,6 +1431,18 @@ void tui_option_set(TUIData *tui, String name, Object value)
   }
 }
 
+void tui_term_unhandled(TUIData *tui, String cmd)
```
Comment (justinmk) [inline src/nvim/tui/tui.c:1434]: doesn't look like this is used anywhere?

File: src/nvim/terminal.c [hunk 2]
```diff
@@ -217,6 +224,8 @@ Terminal *terminal_open(buf_T *buf, TerminalOptions opts)
   vterm_set_utf8(rv->vt, 1);
   // Setup state
   VTermState *state = vterm_obtain_state(rv->vt);
+  // handle custom escape sequences
+  vterm_state_set_unrecognised_fallbacks(state, &vterm_parser_callbacks, rv);
```
Comment (justinmk) [inline src/nvim/terminal.c:228]: wonder what the difference is vs `vterm_screen_set_unrecognised_fallbacks` which is used in https://github.com/neovim/neovim/pull/22159 .
looks like they are equivalent:
https://github.com/neovim/libvterm/blob/9d6d2112335080312ef8c36667fa717ded4f7daf/src/screen.c#L1064-L1067

File: src/nvim/tui/tui.c (hunk 1 above)
Comment (sassanh) [reply src/nvim/tui/tui.c:1434]: Do you think this PR is still relevant considering the changes in the past months? (specially the changes in libvterm)
Also we have [this PR](https://github.com/neovim/neovim/pull/22159) which I think is better approaching this issue.
If this PR is still relevant I'm willing to rebase it and continue working on it.

Comment (justinmk) [reply src/nvim/tui/tui.c:1434]: Yes. Going to try to merge https://github.com/neovim/neovim/pull/22159 today

**Comment by justinmk** [comment]
> > I removed the virtual cursor in tui to show the real cursor.
> 
> That's definitely something we want! #3681

OSC/DSC are now forward to the `TermRequest` autocmd event.

#3681 is still desirable, open a new PR for that.

**Pull request closed without merging**